tools/wptrunner/wptrunner/browsers/content_shell.py

```diff
@@ -0,0 +1,196 @@
+from ..executors import executor_kwargs as base_executor_kwargs
+from ..executors.executorcontentshell import ContentShellRefTestExecutor, \
+    ContentShellCrashtestExecutor, ContentShellTestharnessExecutor
+
+
+__wptrunner__ = {"product": "content_shell",
+                 "check_args": "check_args",
+                 "browser": "ContentShellBrowser",
+                 "executor": {"reftest": "ContentShellRefTestExecutor",
+                     "crashtest": "ContentShellCrashtestExecutor",
+                     "testharness": "ContentShellTestharnessExecutor"},
+                 "browser_kwargs": "browser_kwargs",
+                 "executor_kwargs": "executor_kwargs",
+                 "env_extras": "env_extras",
+                 "env_options": "env_options",
+                 "update_properties": "update_properties",
+                 "timeout_multiplier": "get_timeout_multiplier",}
+
+def check_args(**kwargs):
+    pass
+
+def browser_kwargs(logger, test_type, run_info_data, config, **kwargs):
+    args = kwargs["binary_args"]
+
+    args.append("--ignore-certificate-errors-spki-list=%s" %
+                                  ','.join(chrome_spki_certs.IGNORE_CERTIFICATE_ERRORS_SPKI_LIST))
+
+    webtranport_h3_port = config.ports.get('webtransport-h3')
+    if webtranport_h3_port is not None:
+        args.append(
+            f"--origin-to-force-quic-on=web-platform.test:{webtranport_h3_port[0]}")
+
+    # These flags are specific to content_shell - they activate web test protocol mode.
+    args.append("--run-web-tests")
+    args.append("-") 
+
+    return {"binary": kwargs["binary"],
+            "binary_args": args}
+
+def executor_kwargs(logger, test_type, test_environment, run_info_data,
+                    **kwargs):
+    executor_kwargs = base_executor_kwargs(test_type, test_environment, run_info_data,
+                                           **kwargs)
+    return executor_kwargs
+
+def env_extras(**kwargs):
+    return []
+
+def env_options():
+    return {"server_host": "127.0.0.1",
+            "testharnessreport": "testharnessreport-content-shell.js"}
+
+def update_properties():
+    return (["debug", "os", "processor"], {"os": ["version"], "processor": ["bits"]})
+
+class ContentShellBrowser(Browser):
+    """Class that represents an instance of content_shell.
+
+    Upon startup, the stdout, stderr, and stdin pipes of the underlying content_shell
+    process are connected to multiprocessing Queues so that the runner process can
+    interact with content_shell through its protocol mode.
+    """
+
+    def __init__(self, logger, binary="content_shell", binary_args=[], **kwargs):
+        super().__init__(logger)
+
+        self._args = [binary] + binary_args
+        self._proc = None
+
+    def start(self, group_metadata, **kwargs):
+        self.logger.debug("Starting content shell: %s..." % self._args[0])
+
+        # Unfortunately we need to use the Process class directly because we do not
```

Comment:
Yes, this is very similar to how mozprocess handles output with two distinctions.
1. We use `multiprocessing.Queue` so that other processes can interact with the browser through its ExecutorBrowser. In mozprocess, the IO has to happen in the same process.
2. mozprocess calls `rstrip()` on the lines it gets from stdout before passing it to the output handler. We don't want that, especially when reading binary image data.